### ISSUE TITLE
Hyperlinks in log entries

### ISSUE THREAD

**Issue by lodici** [magarena/magarena#126]
> Cool - I don't know if this is already in the pipeline, but in the log, the referenced card names in brackets (from searching etc) Having them like a hyperlink-card-pop-up, for when the AI 'reveals' a card you're not quite sure of.

**Comment by lodici** [comment]
Start with
```
LogBookViewer.getNewMessagePanel() -> TextLabel.buildComponents()
```

**Comment by lodici** [comment]
I have just committed a load of updates (72e9dd9083c2ab5f7d01a6975e6670f64c926721 to 504e426f1826c1178d02f0ddf54624d2fb5020a9) that represents the first stage in creating interactive card name entries in the log.
![screenshot00299](https://cloud.githubusercontent.com/assets/2443519/9977290/4ad5dc46-5ef9-11e5-9577-e72be4c6404e.png)

The main UI changes are
- the default font for the log and stack is now plain instead of bold so that card names stand out.
- recognised cards (see below) are shown in bold.
- choices are shown in red instead of blue (which I always thought could be mistaken for links).
- currently the card names are not interactive but I should have some more updates shortly which will highlight the location of a card on screen when you mouse over a card name or show you the full sized image if you click on it instead.
- run -DdevMode=true to display the MagicCard ID in the log and as an overlay on the card images.

#### How does it pick out cards from the log message?
The following token is now recognized in a log message.
```
<card_name#id>
```
Currently only messages with ``SN`` (see ``MagicMessage.replaceName()``) will be processed so although this covers quite a lot of messages there will be still be many where the card name will be displayed unbolded. This is the next challenge - the generation of log messages is scattered all over the place, some use ``SN`` replacement, some don't so it would be great if we could come up with a consistent way of of tokenizing the log messages.

**Comment by lodici** [comment]
Second part of this has been committed (fee0b9f223c32b37761fca8a8ebab795b5d34735). Now when you mouse over a card name that is in bold it will highlight its location including the player zone if that is where the card is. Click on the card name to view it full size.

I am still not too sure whether the card highlight is effective. I did experiment with flashing the image which I might still look at before 1.66 release.

**Comment by ShawnieBoy** [comment]
Really liking this - the highlight seems effective enough for me.
This will also require some work on the groovy files which send messages to the log - and also the most important log messages: target creature/target permanent (For those cases where several permanents of the same name are on the battlefield.

**Comment by lodici** [comment]
> also the most important log messages: target creature/target permanent (For those cases where several permanents of the same name are on the battlefield.

Could you give an example or two (a screenshot would be useful).

**Comment by ShawnieBoy** [comment]
These names are essentially generated from MagicTargetChoice.

![pillory](https://cloud.githubusercontent.com/assets/2363180/9990266/d325f3ec-6057-11e5-992d-9bdf663c7c42.png)
MagicPlayAuraEvent is where this name is generated from `targetChoice.getTargetDescription()`

![acceptable](https://cloud.githubusercontent.com/assets/2363180/9990272/e70054ac-6057-11e5-9215-6160547041b9.png)
This covers two links, the random discard as an additional cost, and the target for damage.

These are both scripted with no groovy involved.

**Comment by lodici** [comment]
Thanks, I will take a look.

**Comment by lodici** [comment]
9c6e4cbc9188204ff13cb2db66e3c7ad423941d4 should take care of the first case above (_Pillory of the Sleepless).

**Comment by lodici** [comment]
df355d4a964a206f07e86c833c80215d4415dc7a addresses the second case (random discard).

**Comment by lodici** [comment]
Sorry to be picky but here is the log when Acceptable Losses destroys the target -
![screenshot00313](https://cloud.githubusercontent.com/assets/2443519/10005984/442c725a-60b2-11e5-971e-f1226d24cd01.png)
The full stop should come after the choice (Cylian Elf) I think otherwise it looks like you are duplicating the card name.

**Comment by lodici** [comment]
Here is a different style without the bold font - I did not appreciate how much of it there would be. This is maybe a bit cleaner and easier to read? And blue = link now makes sense.
![screenshot00316](https://cloud.githubusercontent.com/assets/2443519/10006805/375644ec-60b8-11e5-9b9b-3c09ce3a96dc.png)

**Comment by ShawnieBoy** [comment]
Personally I quite liked the look of the bold - Would a blue bold be overdoing it?

**Comment by melvinzhang** [comment]
@lodici Thanks for pointing out the problem with the full stop. Fixed pushed.

I think the non-bold version is easier to read. The coloring of selected choice is a bit inconsistent, currently it is bold blue on the stack but bold red in the log.

**Comment by lodici** [comment]
I will add a button to the log/stack titlebar that will let you cycle through various styles - plain-blue, plain-bold-blue, plain-bold-black, bold-blue. Also will look at updating the stack messages to be consistent with the log.

**Comment by lodici** [comment]
Can you also confirm that the characters, ">",  "<" and "#", should be ok to use as delimiters. I have not really checked, I just assumed that they do not appear in oracle text or card ability descriptions.

**Comment by melvinzhang** [comment]
A quick check with grep should be able to confirm this, by looking for < 
```
grep '".*>.*"' -r release/Magarena/scripts
grep '".*<.*"' -r release/Magarena/scripts
grep '".*#.*"' -r release/Magarena/scripts
grep '<' -r release/Magarena/scripts/*.txt
grep '>' -r release/Magarena/scripts/*.txt
grep '#' -r release/Magarena/scripts/*.txt
```
None of the above returned any results

**Comment by lodici** [comment]
Great. Thanks.

**Comment by lodici** [comment]
I am sure I have asked this before but searching issues has come up with a blank. Can I have an example or test that requires a stack button to be clicked. Thanks.

**Comment by melvinzhang** [comment]
Casting a counter spell requires the user to target a spell on the stack, see src/magic/test/TestCounterFlash.java

**Comment by lodici** [comment]
:+1:

**Comment by lodici** [comment]
I think this is now more or less done, at least for the 1.66 release. I have been running quite a lot of AI vs AI games and using [this debug statement](https://github.com/magarena/magarena/blob/master/src/magic/model/MagicMessage.java#L113) to show card names in log messages that have not yet been tokenized. I have not come across any new ones recently although I can't be 100% certain that all cases have been covered given the complexity of the game.

Tokens are still something that will need to be looked at.

**Comment by melvinzhang** [comment]
@lodici Fantastic work, I think this is a great improvement to keeping track of what is happening in the game. Do update the release README.txt as well, I'm sure our users will be very pleased to learn of this enhancement.

**Comment by lodici** [comment]
Thanks, will update the README before tomorrow.

**Comment by lodici** [comment]
Implemented in 1.66.
